substate-store-impls/src/substate_database_staging.rs

```diff
@@ -0,0 +1,282 @@
+                }
+            }
+            // This particular node key does not exist in the overlay. The substate must be read
+            // from the underlying database.
+            None => OverlayLookupResult::NotFound,
+        };
+
+        match overlay_lookup_result {
+            OverlayLookupResult::Found(substate_value) => substate_value.cloned(),
+            OverlayLookupResult::NotFound => self.root.get_substate(partition_key, sort_key),
+        }
+    }
+
+    fn list_entries_from(
+        &self,
+        partition_key @ DbPartitionKey {
+            node_key,
+            partition_num,
+        }: &DbPartitionKey,
+        from_sort_key: Option<&DbSortKey>,
+    ) -> Box<dyn Iterator<Item = PartitionEntry> + '_> {
+        // This function iterates over entries of the specified partition. Therefore, we don't need
+        // to think about other partitions here. We first check if there are any partition updates
+        // for the specified partition. If there is not, no overlaying is needed and we can just
+        // return the iterator of the root store.
+        match self.overlay.node_updates.get(node_key) {
+            // There is a partition update in the overlay.
+            Some(NodeDatabaseUpdates { partition_updates }) => {
+                match partition_updates.get(partition_num) {
+                    // The partition was reset. None of the substates of this partition that exist
+                    // in the root store "exist" anymore. We just need an iterator over the new
+                    // substates in the reset action.
+                    Some(PartitionDatabaseUpdates::Reset {
+                        new_substate_values,
+                    }) => Box::new(
+                        new_substate_values
+                            .iter()
+                            .map(|(sort_key, substate_value)| {
+                                (sort_key.clone(), substate_value.clone())
+                            })
+                            .sorted(),
+                    ),
+                    // There are some changes that need to be overlayed.
+                    Some(PartitionDatabaseUpdates::Delta { substate_updates }) => {
+                        let underlying = self.root.list_entries_from(partition_key, from_sort_key);
+                        let overlaying = substate_updates
+                            .iter()
+                            .map(|(sort_key, database_update)| match database_update {
+                                DatabaseUpdate::Set(substate_value) => {
+                                    (sort_key.clone(), Some(substate_value.clone()))
+                                }
+                                DatabaseUpdate::Delete => (sort_key.clone(), None),
+                            })
+                            .sorted_by(|(sort_key_a, _), (sort_key_b, _)| {
+                                sort_key_a.cmp(sort_key_b)
+                            });
```

Comment:
> Functional issue: I believe the overlay is missing a filter on from_sort_key here.

Good point on that. I've updated that and added extra tests for this particular case to verify it.

> Possible performance issue: It would be more efficient if we used a structure which saved the overlaid substates in a BTreeMap rather than an IndexMap, as this requires sorting the map constantly.

Yup, agreed that `BTreeMap`s would be better here so that we're not constantly sorting when we do a `list_entries_from`. I've made that switch by introducing a new set of types that all use `BTreeMap` instead of `IndexMap`: `StagingDatabaseUpdates`, `StagingNodeDatabaseUpdates`, and `StagingPartitionDatabaseUpdates`. These types implement `From` and `Into` their original types to ensure that they're always in sync with them and so that we never expose any of these staging types outside this module.